flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala

```diff
@@ -342,21 +368,29 @@ case class Join(
   }
 
   private def testJoinCondition(expression: Expression): Unit = {
-    def checkIfJoinCondition(exp : Expression) = if (exp.children.exists(!_.isInstanceOf[JoinFieldReference])) {
-      failValidation(s"Only join predicates supported. For non-join predicates use Table#where.")
-    }
+    def checkIfJoinCondition(exp : BinaryComparison) =
+      if (exp.children match {
+        case (x : JoinFieldReference) :: (y : JoinFieldReference) :: Nil  =>
+          x.belongsToLeft == y.belongsToLeft
+        case _ => true
+      }) {
+        failValidation(s"Only join predicates supported. For non-join predicates use Table#where.")
+      }
 
     var equiJoinFound = false
-    def validateConditions(exp: Expression) : Unit = exp match {
-      case x: And => x.children.foreach(validateConditions(_))
+    def validateConditions(exp: Expression, isAndBranch: Boolean): Unit = exp match {
```

Comment:
I'm sorry but this check does not catch all cases. For example, it would let `l.a = r.b OR (l.c = r.d && l.e = r.f)` pass.